scipy/cluster/hierarchy.py

```diff
@@ -3392,18 +3514,22 @@ def llf(id):
     >>> hierarchy.set_link_color_palette(None)  # reset to default after use
     >>> plt.show()
 
+    An example with custom leaves ordering:
+
+    >>> from scipy.cluster import hierarchy
+    >>> X = [[0,0], [0,1], [0,4], [2,4], [6,3],[7,3], [8,3]]
+    >>> Z = hierarchy.linkage(X, method = 'single')
+
+    >>> fig, axes = plt.subplots(1, 2, figsize = (8, 3))
+    >>> dn1 = hierarchy.dendrogram(Z, ax = axes[0])
+    >>> dn2 = hierarchy.dendrogram(Z, ax = axes[1], leaves_order = [3,2,1,0,6,4,5])
+    >>> plt.show()
     """
-    # This feature was thought about but never implemented (still useful?):
-    #
-    #         ... = dendrogram(..., leaves_order=None)
-    #
-    #         Plots the leaves in the order specified by a vector of
-    #         original observation indices. If the vector contains duplicates
-    #         or results in a crossing, an exception will be thrown. Passing
-    #         None orders leaf nodes based on the order they appear in the
-    #         pre-order traversal.
-    xp = array_namespace(Z)
+    xp = array_namespace(Z, leaves_order)
     Z = _asarray(Z, order='C', xp=xp)
+    if leaves_order is not None:
+        leaves_order = _asarray(leaves_order, dtype=xp.int64, xp=xp)
+        Z = _reorder_leaves(Z, leaves_order, xp=xp)
```

Comment:
much of the testing done in this PR at the moment probes the "private" function `_reorder_leaves` instead of `dendogram` itself--which means that there are coverage gaps here--these two lines are not hit by tests.

Indeed, I can delete them from this branch locally and the testsuite still passes, which probably isn't desirable since `dendogram` has an arguably unwieldy number of parameters and building up tests that handle their permutations is likely desirable/necessary.